crates/mdk-core/src/image_processing/metadata.rs

```diff
@@ -0,0 +1,384 @@
+                })?;
+
+        metadata.blurhash = generate_blurhash(&img);
+    }
+
+    Ok(metadata)
+}
+
+/// Extract metadata from an already-decoded image
+///
+/// This function extracts dimensions and blurhash from a decoded DynamicImage,
+/// avoiding the need to decode the image again.
+///
+/// # Arguments
+/// * `img` - The decoded image
+/// * `options` - Validation options for dimension checking
+/// * `generate_blurhash_flag` - Whether to generate a blurhash
+///
+/// # Returns
+/// * `ImageMetadata` with dimensions and optional blurhash
+///
+/// # Errors
+/// * `DimensionsTooLarge` / `TooManyPixels` / `ImageMemoryTooLarge` - If dimensions exceed limits
+pub fn extract_metadata_from_decoded_image(
+    img: &image::DynamicImage,
+    options: &ImageValidationOptions,
+    generate_blurhash_flag: bool,
+) -> Result<ImageMetadata, ImageProcessingError> {
+    let width = img.width();
+    let height = img.height();
+
+    // Validate dimensions
+    validate_image_dimensions(width, height, options)?;
+
+    let mut metadata = ImageMetadata {
+        dimensions: Some((width, height)),
+        blurhash: None,
+    };
+
+    // Generate blurhash if requested
+    if generate_blurhash_flag {
+        metadata.blurhash = generate_blurhash(img);
+    }
+
+    Ok(metadata)
+}
+
+/// Generate blurhash for an image
+///
+/// Creates a compact string representation of the image that can be used
+/// to generate a blurred placeholder while the full image loads.
+///
+/// # Arguments
+/// * `img` - The decoded image
+///
+/// # Returns
+/// * `Some(String)` with the blurhash, or `None` if generation fails
+pub fn generate_blurhash(img: &image::DynamicImage) -> Option<String> {
+    // Resize image for blurhash (max 32x32 for performance)
+    let small_img = img.resize(32, 32, image::imageops::FilterType::Lanczos3);
+    let rgb_img = small_img.to_rgb8();
+
+    encode(4, 3, rgb_img.width(), rgb_img.height(), rgb_img.as_raw()).ok()
+}
+
+/// Check if a MIME type is a known safe raster format that supports EXIF sanitization
+///
+/// This function returns true only for raster image formats that:
+/// 1. Can be safely decoded by the `image` crate
+/// 2. Can be re-encoded without loss of format features (e.g., not animated)
+/// 3. Are commonly used formats where EXIF stripping is valuable
+///
+/// Formats explicitly excluded:
+/// - image/gif: May be animated, would be flattened
+/// - image/webp: May be animated, would be flattened
+/// - image/svg+xml: Vector format, cannot be decoded as raster
+/// - Other vector or specialized formats
+pub fn is_safe_raster_format(mime_type: &str) -> bool {
+    matches!(
+        mime_type,
+        "image/jpeg" | "image/png" // Note: GIF and WebP are explicitly excluded because they may be animated
+                                   // and sanitization would flatten them to a single frame.
+                                   // Future work could detect if they're static and handle them.
+    )
+}
```

Comment:
Comment wording is a bit weird here, I'd expect for the "explicitly excluded" ones to be mentioned in the method (to exclude them) but this seems like it's the other way around, there are only a couple of "explicitly included" ones.